### ISSUE TITLE
Inputs do not conform to this spec or are otherwise invalid

### ISSUE THREAD

**Issue by MartinSchere** [Berry-Pool/nami-wallet#71]
I'm getting this error when trying to sign the simplest transaction:

```
cardano-cli transaction build-raw \
    --tx-in 88d9225ff5d46b0efcd56dda20f2ea93b0a5fa232e7f70ab7d626746075a20d1#0 \
    --tx-out addr_test1qzymym7qc47kx65sjhntdl2prq85rj5v8mzv5pvyffnntz4el5zqnamef3yt2kprt866p5f38w6um6642lvzrj80d0xstnnmxq+10 \
    --tx-out addr_test1qzjlc05tyyw264wy7m4u7np5yqdwglks0xhu6765cl4qex9r9kvav4hmznru9px9n7cpa2hmmv4593eegve3t834xppqwskp4t+999823269 \
    --invalid-hereafter 369400 \
    --fee 176721 \
    --out-file tx.raw
```
![ready2](https://user-images.githubusercontent.com/61954243/135723968-0b85860f-af08-483b-880a-0814b6bc49c2.png)

It's an alonzo-era transaction

**Comment by MartinSchere** [comment]
When doing the following:

```
const cborHex =
      "83a4008182582088d9225ff5d46b0efcd56dda20f2ea93b0a5fa232e7f70ab7d626746075a20d10001828258390089b26fc0c57d636a9095e6b6fd41180f41ca8c3ec4ca05844a67358ab9fd0409f7794c48b5582359f5a0d1313bb5cdeb5557d821c8ef6bcd0a82583900a5fc3e8b211cad55c4f6ebcf4c34201ae47ed079afcd7b54c7ea0c98a32d99d656fb14c7c284c59fb01eaafbdb2b42c7394333159e3530421a3b9817a5021a0002b251031a0005a2f89ffff6";

    const rawTx = CardanoWasm.Transaction.from_bytes(
      Buffer.from(cborHex, "hex")
    );
```

It gives the error:

Uncaught (in promise) Deserialization failed in Transaction.witness_set.TransactionWitnessSet because: Invalid cbor: not the right type, expected `Map' byte received `Array'.

**Comment by alessandrokonrad** [comment]
The raw transaction coming from the cardano-cli does not follow the ledger specs actually. That's why Nami is complaining or better said the serialization-lib. You need to transform the cbor code a little bit. Get rid of the empty arrays at the top and replace them with a map and remove the bool in case there is one.

**Comment by MartinSchere** [comment]
OK, so it's possible to build an alonzo-era transaction and sign it with Nami? 

I'm looking for a way to enable smart contracts in my dApp without waiting for the PAB release

**Comment by MartinSchere** [comment]
Yes, given that the PAB will be here in a few weeks, should the link handler be implemented in Nami? Or the dApp connector is made with the serialization lib?

**Comment by MartinSchere** [comment]
What is the bool thing about?

**Comment by alessandrokonrad** [comment]
> What is the bool thing about?

I think it's an extra flag that tells if the contract evaluates to true or false. I don't know exactly the purpose of it.

**Comment by MartinSchere** [comment]
I created a `toSpec` helper function and it worked:

```js
  public toSpec(txFilePath: string): string {
    const file = JSON.parse(readFileSync(txFilePath, 'utf-8'));

    const { cborHex } = file;
    const decoded = cbor.decode(cborHex);
    const newCborDecoded = [];

    for (let i of decoded) {
      if (Array.isArray(i)) {
        newCborDecoded.push(new Map());
      } else if (typeof i !== 'boolean') {
        newCborDecoded.push(i);
      }
    }
    const result = cbor.encode(newCborDecoded).toString('hex');
    return result;
  }
```

**Comment by ElGatoLoco** [comment]
> I created a `toSpec` helper function and it worked:
> 
> ```js
>   public toSpec(txFilePath: string): string {
>     const file = JSON.parse(readFileSync(txFilePath, 'utf-8'));
> 
>     const { cborHex } = file;
>     const decoded = cbor.decode(cborHex);
>     const newCborDecoded = [];
> 
>     for (let i of decoded) {
>       if (Array.isArray(i)) {
>         newCborDecoded.push(new Map());
>       } else if (typeof i !== 'boolean') {
>         newCborDecoded.push(i);
>       }
>     }
>     const result = cbor.encode(newCborDecoded).toString('hex');
>     return result;
>   }
> ```

This is all good until you actually have some data in those arrays that needs to be transformed. Did you by any chance come up with a way to transform arrays to maps properly?
